truffle/src/com.oracle.truffle.host/src/com/oracle/truffle/host/HostLibrary.java

```diff
@@ -0,0 +1,25 @@
+package com.oracle.truffle.host;
+
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.library.GenerateLibrary;
+import com.oracle.truffle.api.library.Library;
+import com.oracle.truffle.api.library.LibraryFactory;
+
+@GenerateLibrary(receiverType = HostInstance.class)
+abstract class HostLibrary extends Library {
+    private static final LibraryFactory<HostLibrary> FACTORY = HostLibraryGen.resolve(HostLibrary.class);
+    private static final HostLibrary UNCACHED = FACTORY.getUncached();
+
+    static HostLibrary getUncached() {
+        CompilerAsserts.neverPartOfCompilation("HostLibrary needs to be fast on fast-path");
+        return UNCACHED;
+    }
+    
+    static LibraryFactory<HostLibrary> getFactory() {
+        return FACTORY;
+    }
+    
+    public HostInstance asHostInstance(Object obj) {
+        return obj instanceof HostInstance ? (HostInstance) obj : null;
```

Comment:
Interface type checks are very expensive. Any reason why there is not a isHostInstance and asHostInstance. 
The first returns true|false whether it is a host object. And the second returns not the HostObject wrapper but the underlying real host object. asHostInstance then could fail with an UnsupportedMessageException if it is used with the wrong implementation.